content/en/blog/_posts/2020-09-02-scaling-kubernetes-networking-endpointslices.md

```diff
@@ -0,0 +1,47 @@
+
+Consider an example where a Service is backed by 15 pods. We'd end up with a single Endpoints resource that tracked all of them. If EndpointSlices were configured to store 5 endpoints each, we'd end up with 3 different EndpointSlices:
+![EndpointSlices](/images/blog/2020-09-02-scaling-kubernetes-networking-endpointslices/endpoint-slices.png)
+
+By default, EndpointSlices will store as many as 100 endpoints each, though this can be configured with the --max-endpoints-per-slice flag on kube-controller-manager.
+
+## EndpointSlices provide 10x scalability improvements
+This API dramatically improves networking scalability. Now when a Pod is added or removed, only 1 small EndpointSlice needs to be updated. This difference becomes quite noticeable when hundreds or thousands of Pods are backing a single Service. 
+
+Potentially more significant, now that all Pod IPs for a Service don't need to be stored in a single resource, we don't have to worry about the size limit for objects stored in etcd. EndpointSlices have already been used to scale Services beyond 100,000 network endpoints.
+
+All of this is brought together with some significant performance improvements that have been made in kube-proxy. When using EndpointSlices at scale, significantly less data will be transferred for endpoints updates and kube-proxy should be faster to update iptables or ipvs rules. Beyond that, Services can now scale at least 10x beyond any previous limitations.
+
+## EndpointSlices enable new functionality
+As a new API, EndpointSlices were built to enable some exciting new functionality in future Kubernetes releases. This will include dual-stack Services, topology aware routing, and endpoint subsetting.
+
+Dual-Stack Services are an exciting new feature that has been in development alongside EndpointSlices. They will utilize both IPv4 and IPv6 addresses for Services and rely on the addressType field on EndpointSlices to track these addresses by IP family. 
+
+Topology aware routing will update kube-proxy to prefer routing requests within the same zone or region. This makes use of the topology fields stored for each endpoint in an EndpointSlice. As a further refinement of that, we're exploring the potential of endpoint subsetting. This would allow kube-proxy to only watch a subset of EndpointSlices. For example, this might be combined with topology aware routing so that kube-proxy would only need to watch EndpointSlices containing endpoints within the same zone. This would provide another very significant scalability improvement.
+
+## What does this mean for the Endpoints API?
+Although the EndpointSlice API is providing a newer and more scalable alternative to the Endpoints API, the Endpoints API will continue to be considered generally available and stable. The most significant change planned for the Endpoints API will involve beginning to truncate Endpoints that would otherwise run into scalability issues.
```

Comment:
Not sure what the deprecation policy is here is but I think a stable API like Endpoints might be considered stable throughout the 1.x.x major version series? If so, let's make that clear.

Reply:
Good catch. I'd thought that all GA APIs are guaranteed to be stable for all 1.x releases, but can't find anything in the docs directly stating that. The official [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) seems to indicate that a GA API can be removed as early as 1 year after deprecation. I'm unaware of any plans to deprecate the Endpoints API though and do not think it will happen. I'm just hesitant to commit to something not happening since I don't actually control it.

Maybe @johnbelamaric has some guidance here on what we can/can't say about the Endpoints API. My understanding is that it will continue to be supported indefinitely. As a GA API can we guarantee that it will continue to be supported for all 1.x releases?

Reply:
We don't really follow [semver](https://semver.org/), it theoretically could be removed according to the deprecation policy that Rob links.

Reply:
any suggestions for edit or shall we leave it like it is?

Reply:
I can't think of a great way to improve this. Mentioning that "we currently have no plans to deprecate Endpoints" may actually result in readers becoming concerned about future deprecation of this API. On the other we may regret a stronger statement if a year from now we end up deprecating the API. (I don't think that will happen, it's just always dangerous to commit to something never happening).

Reply:
> API elements may only be removed by incrementing the version of the API group.

So you might find that Endpoints v1 gets removed but I presume only in favor of Endpoints v2.

Is there any way _ever_ to remove Endpoints? Sounds like that might need a KEP!

Reply:
Yeah that's where I got confused as far as the deprecation policy as well. It seemed to only cover deprecating a GA API if a replacement/newer version existed. I'm not sure if EndpointSlices would qualify as a direct replacement or not.